R/groups.R

```diff
@@ -0,0 +1,190 @@
+#'   * `username`: The user's username
+#'   * `first_name`: The user's first name
+#'   * `last_name`: The user's last name
+#'   * `user_role`: The user's role. It may have a value of administrator,
+#'     publisher or viewer.
+#'   * `created_time`: The timestamp (in RFC3339 format) when the user
+#'     was created in the Posit Connect server
+#'   * `updated_time`: The timestamp (in RFC3339 format) when the user
+#'     was last updated in the Posit Connect server
+#'   * `active_time`: The timestamp (in RFC3339 format) when the user
+#'     was last active on the Posit Connect server
+#'   * `confirmed`: When false, the created user must confirm their
+#'     account through an email. This feature is unique to password
+#'     authentication.
+#'   * `locked`: Whether or not the user is locked
+#'   * `guid`: The user's GUID, or unique identifier, in UUID RFC4122 format
+#'
+#' @details
+#' Please see https://docs.posit.co/connect/api/#get-/v1/groups/-group_guid-/members
+#' for more information.
+#'
+#' @examples
+#' \dontrun{
+#' library(connectapi)
+#' client <- connect()
+#'
+#' # get the first 20 groups
+#' groups <- get_groups(client)
+#'
+#' group_guid <- groups$guid[1]
+#'
+#' get_group_members(client, guid = group_guid)
+#' }
+#'
+#' @family groups functions
+#' @export
+get_group_members <- function(src, guid) {
+  validate_R6_class(src, "Connect")
+
+  res <- src$group_members(guid)
+
+  out <- parse_connectapi(res$results)
+
+  return(out)
+}
+
+#' Get content access permissions for a group or groups
+#'
+#' @param src A Connect client object
+#' @param guids A character vector of group guids
+#'
+#' @return
+#' A tibble with the following columns:
+
+#'   * `group_guid`: The group's GUID
+#'   * `group_name`: The group's name
+#'   * `content_guid`: The content item's GUID
+#'   * `content_name`: The content item's name
+#'   * `content_title`: The content item's title
+#'   * `access_type`: The access type of the content item ("all", "logged_in", or "acl")
+#'   * `role`: The access type that members of the group have to the
+#'     content item, "publisher" or "viewer".
+#'
+#' @examples
+#' \dontrun{
+#' library(connectapi)
+#' client <- connect()
+#'
+#' # Get a data frame of groups
+#' groups <- get_groups(client)
+#'
+#' # Get permissions for a single group by passing in the corresponding row.
+#' get_group_content(client, groups[1, "guid"])
+#' dplyr::filter(groups, name = "research_scientists") %>%
+#'   dplyr::pull(guid) %>%
+#'   get_group_content(client, .)
+#'
+#' # Get permissions for all groups by passing in all group guids.
+#' get_group_content(client, groups$guid)
+#' }
+#'
+#' @family groups functions
+#' @export
+#' @importFrom rlang .data
+get_group_content <- function(src, guids) {
+  validate_R6_class(src, "Connect")
+
+  purrr::map_dfr(
+    guids,
+    ~ get_group_content_impl(src = src, guid = .x)
+  )
+}
+
+#' @importFrom rlang .data
+get_group_content_impl <- function(src, guid) {
```

Comment:
Do we still need this separate level of nested functions here still? I think it made more sense when there were a few more branches, but it feels unnecessary here. 

I also wonder if we can reduce the number of (practical) loops we're going through. We call this in a `purrr::map..` up above, and then there is at least one `purrr::map...` below which includes two anonymous functions which also practically go through each element checking for equality. These might all be necessary, but when I was trying to simplify this found the levels of nesting to be not-totally-obvious what was doing what where.

Reply:
Hmm, I'm not sure. I found the distinction at this level to be helpful when thinking about the code — thinking about parsing each group's content into a data frame and then mapping that across `map_dfr` was helpful. But it would be possible to put this in an anonymous function call.

It isn't immediately obvious to me how to reduce the level of nestedness.

- This level is necessary, because we receive multiple GUIDs, and each one requires some level of processing.
- The inner processing is fairly nested because only the top level of the incoming JSON is parsed into a data frame, and the API returns a fairly nested structure for `permissions` (see below).

Reply:
> This level is necessary, because we receive multiple GUIDs, and each one requires some level of processing.

Maybe we could make the name of this function make this clearer? If it were something like `get_one_group_content` that makes it a bit clearer that it's getting content for a single guid?